pkg/util/interceptor/server_id_interceptor.go

```diff
@@ -0,0 +1,95 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interceptor
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+
+	"github.com/milvus-io/milvus/pkg/util/merr"
+	"github.com/milvus-io/milvus/pkg/util/paramtable"
+)
+
+const ServerIDKey = "ServerID"
+
+// ServerIDValidationUnaryServerInterceptor returns a new unary server interceptor that
+// verifies whether the target server ID of request matches with the server's ID and rejects it accordingly.
+func ServerIDValidationUnaryServerInterceptor() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		md, ok := metadata.FromIncomingContext(ctx)
+		if !ok {
+			return handler(ctx, req)
+		}
+		values := md.Get(ServerIDKey)
+		if len(values) == 0 {
+			return handler(ctx, req)
+		}
+		serverID, err := strconv.ParseInt(values[0], 10, 64)
+		if err != nil {
+			return handler(ctx, req)
+		}
+		if serverID != paramtable.GetNodeID() {
```

Comment:
Prefer to pass nodeID instead of paramtable.GetNodeID() as it may be incorrect in integration test

Reply:
The session is not initialized when starting the grpc loop (which invokes the interceptor), thus we cannot pass the nodeID into the interceptor.